### DISCUSSION TITLE
C++의 가변 길이 배열에 대한 질문

### DISCUSSION THREAD

**Discussion by minjae9610** [SSAFY-10th/algorithm#11] [Q&A] [answered]
기존에 C를 학습할때는 C90의 POSIX 표준 환경에서 학습을 진행했어서 C++의 몇가지 개념들이 헷갈립니다.

C++에서 가변 길이 배열에 대해 알아보던 중, 벡터 컨테이너와 동적 할당으로 만들어진 배열과 어떠한 차이가 있는것인지 명확하게 이해가 가지 않습니다.

[GCC 문서](https://gcc.gnu.org/onlinedocs/gcc/Variable-Length.html)에 따르면 C++의 표준이 아닌 C99의 확장 문법과 C11의 선택 구현사항이 적용되는것으로 이해했습니다.

- 이러한 가변 길이 배열의 경우에는 컴파일 단에서 길이의 정보를 가지고 있지 않게 되고, 이렇게 되면 해당 배열을 메모리의 스택 영역에 할당이 불가능한것이 아닌가요?
- 만약 heap에 할당이 되게 된다면 동적할당과의 차이는 무엇인지, 또한 동적할당과 같이 힙 메모리가 부족하거나, 파이프라인의 문제로 인해 할당 실패가 발생할 수 있는지 궁금합니다.
- 또한, 사용 후 메모리의 회수는 어떻게 처리되는지도 궁금합니다.

해당 부분에 대한 설명이나 참고할 만한 자료가 있다면 부탁드립니다.

**Comment by gabalja** [comment]
1번은 아는게 없고요..
2번에 관해서는 일단 힙에 할당이 되는 것으로 알고 있었는데 찾아보니 capacity라는 걸 자체적으로 쓰는 것 같습니다. (https://aaamy91.tistory.com/58)
3번은 clear를 사용합니다. (https://d-yong.tistory.com/74) 
제가 알고있는건 이정도입니다 ㅠ

**Reply by minjae9610** [reply]
앗 둘다 Vector 에 대한 이야기네요

제가 궁금한건 VLA(가변 길이 배열)에 대한 이야기였습니다.
권장되지 않는다는것은 알고 있지만 왜 권장되지 않는지 작동 방식이 궁금해져서 그렇습니다.

**Comment by minjae9610** [comment]
일단 Stack에 저장된다는 사실은 찾았습니다. 또한 권장되지 않는 방법이라는것도요.

그렇다면 컴파일단에서 함수 Stack의 공간을 어떻게 예측하는지 감이 오지 않습니다. 해당 부분에 대해서 추가적으로 궁금합니다.

**Comment by minjae9610** [comment, marked as answer]
찾아 본 결과 Stack에 고정크기로 공간을 잡아두고 사용하는 방식이라고 합니다. 실제로 동적 할당이 되지 않는다고 하네요. :(
따라서 메모리가 효율적이지도 못하고, 스택 오버플로우의 문제가 발생할 가능성이 있다고 합니다.
이러한 문제 때문에 필수 구현사항에서 선택 구현 사항으로 변경되었고, 사용이 지양되는 문법이라고 보면 될 듯 합니다.
